Fixed #22804 -- Warn on unsafe value of 'sep=' in Signer

jaap3 opened the pull request.

timgraham commented on django/core/signing.py:
```diff
@@ -151,6 +155,14 @@ class Signer(object):
         self.sep = force_str(sep)
+        sep_is_invalid = sep != ":" and (
+            not self.sep or
+            self.sep[0] in _SEP_UNSAFE or
```
why checking only the first and last values of `self.sep`? how about `any(sep in _SEP_UNSAFE for sep in self.sep)`?

timgraham commented on django/core/signing.py:
```diff
@@ -38,15 +38,19 @@
 from django.conf import settings
 from django.utils import baseconv
 from django.utils.crypto import constant_time_compare, salted_hmac
+from django.utils.deprecation import RemovedInNextVersionWarning
```
should use `RemovedInDjango110Warning`

timgraham commented:
See [Deprecating a feature](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature) for documentation requirements.

jaap3 commented:
I've updated the pull request. Changed the validity test to a regex based one. The separator should not be indistinguishable from the signature, so as long as it contains a character outside the range of urlsafe characters it should be fine.

Updated the docs. There already is a note about the separator on https://docs.djangoproject.com/en/1.8/topics/signing/#django.core.signing.Signer so I didn't expand on that.

timgraham commented on django/core/signing.py:
```diff
@@ -150,11 +155,23 @@ class Signer(object):
+    def sep(self, value):
+        sep = force_str(value)
+        if _SEP_UNSAFE.match(sep):
+            warnings.warn('Unsafe Signer separator: %r (cannot be empty or consist of only A-z0-9-_=)' % sep,
```
are you sure about this point of "consist of only A-z0-9-_=", as opposed to those characters may not appear in the separator? (not too familiar with this restriction myself)
I don't know that the complexity of a `setter` is necessary. I'd just modify the test to initialize a `Signer` each time.

jaap3 replied:
sep is used in as an argument to `rsplit` in `unsign` like so:

    value, sig = signed_value.rsplit(self.sep, 1)

So as long as the separator contains at least one character that will never appear in the signature string it should be safe.

I can remove the `setter`. My reasoning behind it is that `Signer().sep` is (undocumented, but) public, so it should be checked at all times. Same could be said for key and salt. I do agree that it bloats the code a fair bit. So I guess it's your call.

timgraham replied:
I agree if we wanted to add the `setter`, we should have it for `salt` as well and I don't see much benefit to the extra complexity.

timgraham commented on tests/signing/tests.py:
```diff
@@ -112,6 +113,26 @@ def test_works_with_non_ascii_keys(self):
+
+    def test_invalid_sep(self):
+        """should warn on invalid separator"""
+        separators = ["", "-", "abc",]
```
please use single quotes for consistency

timgraham commented:
doc edits: http://dpaste.com/20NM7T4 (please wrap at 79 chars)

jaap3 commented:
Updated the PR

timgraham commented:
merged in 0d71349773f8d2e8160d76b5c91a1896ece772db, thanks!
